File: python/paddle/linalg.py
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from paddle import Tensor

from .tensor import inverse as inv
from .tensor.linalg import (
    cholesky,
    cholesky_inverse,
    cholesky_solve,
    cond,
    corrcoef,
    cov,
    det,
    diagonal,
    eig,
    eigh,
    eigvals,
    eigvalsh,
    fp8_fp8_half_gemm_fused,
    householder_product,
    lstsq,
    lu,
    lu_solve,
    lu_unpack,
    matmul,
    matrix_exp,
    matrix_norm,
    matrix_power,
    matrix_rank,
    matrix_transpose,
    multi_dot,
    norm,
    ormqr,
    pca_lowrank,
    pinv,
    qr,
    slogdet,
    solve,
    svd,
    svd_lowrank,
    svdvals,
    triangular_solve,
    vecdot,
    vector_norm,
)

__all__ = [
    'cholesky',
    'cholesky_inverse',
    'norm',
    'matrix_norm',
    'vecdot',
    'vector_norm',
    'cond',
    'cov',
    'corrcoef',
    'cross',
    'inv',
    'eig',
    'eigvals',
    'multi_dot',
    'matrix_rank',
    'matrix_transpose',
    'matmul',
    'svd',
    'svdvals',
    'qr',
    'householder_product',
    'pca_lowrank',
    'svd_lowrank',
    'lu',
    'lu_solve',
    'lu_unpack',
    'matrix_exp',
    'matrix_power',
    'det',
    'slogdet',
    'eigh',
    'eigvalsh',
    'pinv',
    'solve',
    'cholesky_solve',
    'triangular_solve',
    'lstsq',
    'ormqr',
    'fp8_fp8_half_gemm_fused',
    'diagonal',
]


def cross(
    input: Tensor,
    other: Tensor,
    *,
    dim: int = -1,
    out: Tensor | None = None,
) -> Tensor:
    """
    Computes the cross product of two 3-dimensional vectors along the specified dimension.

    Refer to :ref:`api_paddle_cross` for more detail.

    Args:
        input (Tensor): The first input tensor.
        other (Tensor): The second input tensor.
        dim (int, optional): The dimension along which to compute the cross product.
            Default: -1, which means the last dimension.
        out (Tensor, optional): The output tensor. Default: None.
        name (str, optional): Name for the operation. For more information, please
            refer to :ref:`api_guide_Name`. Default: None.

    Returns:
        Tensor: The cross product of ``input`` and ``other``.
    """
    import paddle

    return paddle.cross(input, other, axis=dim, out=out)

File: paddle/tcmpt/hapi/lib/linalg.cc
/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/tcmpt/hapi/include/linalg.h"

#include <memory>

#include "glog/logging.h"

#include "paddle/tcmpt/api/include/core.h"
#include "paddle/tcmpt/api/include/infershape.h"
#include "paddle/tcmpt/core/convert_utils.h"
#include "paddle/tcmpt/core/dense_tensor.h"
#include "paddle/tcmpt/core/kernel_context.h"
#include "paddle/tcmpt/hapi/lib/kernel_dispatch.h"
#include "paddle/tcmpt/infershape/binary.h"

namespace paddle {
namespace experimental {

Tensor dot(const Tensor& x, const Tensor& y) {
  // 1. Get kernel signature and kernel
  auto kernel_key_set = ParseKernelKeyByInputArgs(x);
  auto kernel_key = kernel_key_set.GetHigestPriorityKernelKey();
  auto kernel =
      pt::KernelFactory::Instance().SelectKernelOrThrowError("dot", kernel_key);

  // 2. Get Device Context
  auto* dev_ctx = GetDeviceContextByBackend(kernel_key.backend());
  auto kernel_context = pt::KernelContext(*dev_ctx);

  // 3. Auto data transform
  auto dense_x = std::dynamic_pointer_cast<pt::DenseTensor>(x.impl());
  kernel_context.EmplaceBackInput(dense_x);
  auto dense_y = std::dynamic_pointer_cast<pt::DenseTensor>(y.impl());
  kernel_context.EmplaceBackInput(dense_y);
  // TODO(chenweihang): add transform impl

  // 4. InferShape
  auto out_meta = DotInferShape(dense_x->meta(), dense_y->meta());

  // 5. Prepare outputs
  Tensor out;
  auto dense_out =
      std::make_shared<pt::DenseTensor>(out_meta, pt::TensorStatus());
  kernel_context.EmplaceBackOutput(dense_out);
  out.set_impl(dense_out);
  out.set_backend_set(x.backend_set());

  // 6. Call kernel
  kernel(&kernel_context);

  return out;
}

}  // namespace experimental
}  // namespace paddle
